incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1880,36 +1879,55 @@ private void onDecodePublish(
-                    publishBuilder.format(f -> f.set(mqttPublishHeaderRO.payloadFormat));
-                    publishBuilder.responseTopic(mqttPublishHeaderRO.responseTopic);
-                    publishBuilder.correlation(c -> c.bytes(mqttPublishHeaderRO.correlationData));
-                    final Array32FW<MqttUserPropertyFW> userProperties = userPropertiesRW.build();
-                    userProperties.forEach(c -> publishBuilder.propertiesItem(p -> p.key(c.key()).value(c.value())));
-                });
-
+            int reasonCode = SUCCESS;
+            if (mqttPublishHeaderRO.qos > maximumQos)
+            {
+                reasonCode = QOS_NOT_SUPPORTED;
+            }
+            else if (mqttPublishHeaderRO.retained && !retainAvailable(capabilities))
+            {
+                reasonCode = RETAIN_NOT_SUPPORTED;
+            }
 
-            final MqttDataExFW dataEx = builder.build();
-            if (stream != null)
+            if (reasonCode != SUCCESS)
             {
-                stream.doPublishData(traceId, reserved, payload, dataEx);
+                onDecodeError(traceId, authorization, reasonCode);
+                decoder = decodeIgnoreAll;
+            }
+            else
+            {
+                final int topicKey = topicKey(mqttPublishHeaderRO.topic);
+                MqttPublishStream stream = publishStreams.get(topicKey);
+
+                final MqttDataExFW.Builder builder = mqttPublishDataExRW.wrap(dataExtBuffer, 0, dataExtBuffer.capacity())
+                    .typeId(mqttTypeId)
+                    .publish(publishBuilder ->
+                    {
+                        publishBuilder.qos(mqttPublishHeaderRO.qos);
+                        publishBuilder.flags(mqttPublishHeaderRO.flags);
+                        publishBuilder.expiryInterval(mqttPublishHeaderRO.expiryInterval);
+                        publishBuilder.contentType(mqttPublishHeaderRO.contentType);
+                        publishBuilder.format(f -> f.set(mqttPublishHeaderRO.payloadFormat));
+                        publishBuilder.responseTopic(mqttPublishHeaderRO.responseTopic);
+                        publishBuilder.correlation(c -> c.bytes(mqttPublishHeaderRO.correlationData));
+                        final Array32FW<MqttUserPropertyFW> userProperties = userPropertiesRW.build();
+                        userProperties.forEach(c -> publishBuilder.propertiesItem(p -> p.key(c.key()).value(c.value())));
+                    });
+
+
+                final MqttDataExFW dataEx = builder.build();
+                if (stream != null)
+                {
+                    stream.doPublishData(traceId, reserved, payload, dataEx);
+                }
+                doSignalKeepAliveTimeout();
             }
-            doSignalKeepAliveTimeout();
         }
 
-        private void onDecodeSubscribe(
+        private int onDecodeSubscribe(
             long traceId,
             long authorization,
+            int progress,
```

Comment:
What is triggering this change?

Reply:
We don't want to progress decoding subscribes until we have the capabilities, as we don't know if we need to refuse wildcard/shared subscriptions etc.

Reply:
So we want to buffer the subscribe in the decodebuffer until we receive the capabilitie